### PR TITLE
[JENKINS-32936] Support for multiple layouts in Jelly Layout tag

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-32936](https://issues.jenkins-ci.org/browse/JENKINS-32936) and [JENKINS-25488](https://issues.jenkins-ci.org/browse/JENKINS-25488)

Basically, this PR allows to use `<l:layout type="[two-column or one-column]">` in order to use a full-width size layout (main-content) or a two-column layout (left-side + main-content).

This PR is a follow up of #2029 but in this case agaisnt `master` (now 2.x):
- [x] Replace the solution introduced [here](https://github.com/jenkinsci/jenkins/commit/2cc9853ad664fe0c73ab0643b891603529de5e72)
- [x] Verify the solution proposes with three views: New item (`newJob.jelly`), Job configuration (`configuration.jelly`) and About (`AboutJenkins/index.jelly`)

Tested on Firefox and Chrome on Linux and Mac. Windows environment are welcome.

Part of these modifications were commented and reviewed by @KostyaSha, @kzantow, @tfennelly and @daniel-beck in its previous [PR](https://github.com/jenkinsci/jenkins/pull/2029)

Following some screenshots (1024x800px):

captura el 2016-04-12 a las 12 48 12

captura el 2016-04-12 a las 12 50 38

captura el 2016-04-12 a las 12 50 55

captura el 2016-04-12 a las 12 52 31

@reviewbybees

For reviewers [this](https://github.com/jenkinsci/jenkins/pull/2253/files?w=1)

### CODE REVIEW

File: core/src/main/resources/hudson/AboutJenkins/index.jelly [hunk 1]
```diff
@@ -25,44 +25,47 @@ THE SOFTWARE.
...
 
-      <p>
-        ${%dependencies}
-      </p>
-      <j:set var="uri" value="${it.licensesURL}"/>
-      <j:choose>
-          <j:when test="${uri != null}">
+  <l:layout norefresh="true" type="one-column" title="${%about(app.VERSION)}">
+
+    <l:main-panel>
```
Comment (kzantow) [inline core/src/main/resources/hudson/AboutJenkins/index.jelly]: Maybe to simplify this PR, only limit it to the layout change and specific redos for the other approach?

Comment (recena) [inline core/src/main/resources/hudson/AboutJenkins/index.jelly]: I tried to do that but when you add DOM modifications is easy to reach out a lot of indentation changes, etc... The approach is the same than https://github.com/jenkinsci/jenkins/pull/2029

**Comment by recena** [comment]
There is another use case that new different layouts: [JENKINS-33498](https://issues.jenkins-ci.org/browse/JENKINS-33498)

**Comment by amuniz** [comment]
:bee: (as far as I got it)

**Comment by recena** [comment]
@reviewbybees done

**Comment by gusreiber** [comment]
I think this PR is somewhat wrong headed in its approach. As I see it, we actually don't want the be continuing a UI path that is based on situational layout results rather than logical grouping of components. What is important is that certain chunks of data and configuration be available to the user to see on the page or suppress on the page not weather or not that particular chunk of data is "left column data" vs. "right column data". Depending on the geometry of the client device, having columns at all may or may not make any sense.

Taking the item create and item configuration pages as examples, I made the somewhat rushed (though I think correct) decision, that in most contexts, for most users, the 'context actions list', "build queue", and ''executor status" blocks of data are given too much prominence on those pages. But ideally, the user should be able to make that decision for her/himself.

I would much rather see Jenkins generating logical markup that made no reference to layout particulars than more reference to layout particulars. Ideally that logical abstraction would be carried as deep into the stack as possible.

...then, future theme authors can manipulate item placement with clear knowledge of what the pieces are that they are manipulating, rather than needing to reverse engineer based on assumptions about what it meant in a previous layout to have things placed in the "left vs. right" column bucket.

🐜 <--- I am not sure I know what the ant means, but I hope you will take this concern of advancing logical structure over layout declarations and not rush to open a new arbitrary junk drawer into which plugin authors might stuff more junk.

**Comment by recena** [comment]
@gusreiber I'm not sure if we are speaking the same language, I means terms, concepts, approach, etc...

This PR only tries to provide a homogenous and stable solution to something it has not been resolved in Jenkins for a long time: remove the sidepanel (left menu) when no needed or does not make sense.

It is simpe.

In fact, Jenkins 2.0 has needed it and the solution has passed to hack the layout. There are some examples in plugins diong the same. This involves a fragile situation due to CSS files are not modified because can broken plugins.

As I see, the right place is `<layout />` Jelly tag.

This PR is not new, it is only a follow up of https://github.com/jenkinsci/jenkins/pull/2029 seems he received possitive feedback from: @amuniz, @KostyaSha, @tfennelly, @kzantow

**Comment by recena** [comment]
> Depending on the geometry of the client device, having columns at all may or may not make any sense.

There are techniques to walk through requirements.

Obviously, this is not a final and definitive solution but it was something easy to do and solves a problem. I agree with using a grid layout like [this](https://github.com/jenkinsci/jenkins/blob/master/war/src/main/webapp/css/responsive-grid.css) and avoiding partial solutions like you can find in the page header of Jenkins.

**Comment by recena** [comment]
I think [this](https://smacss.com/book/) could be a good starting point for Jenkins 2.0.

**Comment by KostyaSha** [comment]
👍

**Comment by gusreiber** [comment]
@recena sure, more tools in the toolbox in a sense doesn't hurt anything. To the extent pages in Jenkins are feeling the need to declare, "I am a 2 column layout" vs. "I am a 1 column layout", this is a cleaner means of making that declaration that what exists today. My concern is that in making this option cleaner and clearer, you are essentially making a promise to future plugin authors that they can decide to have 1 column vs. 2 columns and implicit in that promise is that the stuff in those columns isn't going to change on them.

I think this is a particularly bad time to be strengthening promises about page layout particulars.

If we were going to expand the amount of layout control we were offering Jenkins page authors, I would prefer if it looks more like this: 
    <l:layout includedResources="[buildQueue,contenxtActions,execStatus]">

Then, the sort of layout location and needs for say the Build Queue can be determined by the needs of the Build Queue. I suppose these attributes don't have to be exclusive, but even still, I would rather see the layout tag augmented with the attribute "short-content-form" vs. "long-content-form", rather than the more results specific attribute names "one-column" vs. "two-columns".

...that is my concern, do with it as you will.

**Comment by gusreiber** [comment]
...then a layout template author can know what content the template will need to support, the components being included might also have a chance to influence their rendering, and different layout templates could be swapped around to produce different layout results. Things like column counts might be conditional in fancy layout templates, strictly determined in less fancy templates, or maybe non-applicable in other cases.

**Comment by KostyaSha** [comment]
@gusreiber could you propose PR?

**Comment by gusreiber** [comment]
@KostyaSha At this moment, no. But if what I am saying makes sense and we agree this is a problem, yes. Thus, I am raising the concern now, that if we invite Jenkins developers to start dictating layout particulars from outside the layout tag (a column format, 1 vs 2 vs 3 columns, with a presumed but not declared certain set of stuff in the columns) rather than passing requirement information to the layout tag (show these 7 things on the page, except the page to be crowded, tune it for this use case) we are even more locking ourselves into support of layout particulars we may very soon regret.

...I think things like layout tags ought to maintain the responsibility of resolving the screen geometry in the most optimal way for the particulars of the content it serves. If you explicitly pass in that the layout should be 2 columns, the layout tag can't really do its primary role. 2 columns _is_ a very particular page geometry solution. Then the layout tag is reduced to a junk-box for including things.

...if later we decide we don't really like breaking the world up into columns, we will need to go back and look at all the pages where the layout tag is used, figure out what was intended by the "type" attribute, and decide if the 1 vs 2 columns setting can be safely ignored.

**Comment by recena** [comment]
@gusreiber 

1. As I imagined your concept of layout and mine, it seems different.
1. The solution proposed works fine without *ANY* changes in the plugins.
1. Only pages in plugins or Jenkins core pages that need full-width size will need to add this attribute `type="one-column"`.
1. This propose does not making a promise to future plugin developers, no more than it already does with `<layout />` Jelly tag,

> I think this is a particularly bad time to be strengthening promises about page layout particulars.

In Jenkins always seems a bad moment, specially when we are talking about doing changes in the UI.

Jenkins 2.0 has introduced a _pseudo_ one-column layout (new item, configuration job, etc.) BUT it is not reusable for plugins and is not extensible. IMO a clear example of hack, _modify something to reach what I need without thinking on third-parties and side effects_.

To avoid another _three-hundred_ [comments PR](https://github.com/jenkinsci/jenkins/pull/2117) I propose to wait for more feedback from the community.

**Comment by recena** [comment]
Probably if Jenkins had something like that http://www.oracle.com/webfolder/ux/middleware/alta/patterns/PageDesign.html we could improve some aspects that are being claimed by some users, companies, etc...

This PR goes in that direction. Standardize the UI layout.

**Comment by gusreiber** [comment]
@recena I would agree that the "Alta" model that you are showing is the good version of what Jenkins does a bad job of today. Doing a better job of that would be better, sure... That left column/main column layout made perfect sense when the assumption was that all apps ought to look something like the windows explorer or MS Outlook. But even email no longer looks like this sidebar layout convention any more:
screen shot 2016-04-27 at 2 06 57 pm

...that there is a "left column" and that column is "navigation area" is an outdated idea. So going in the "Alta" direction of doing a good job of going in the wrong direction is still going in the wrong direction.

In the new google mail example, notice you have different organizational metaphors determined by the sorts of information being organized. I would like to have a world where theme authors could simply change the `<layout>` template and move organizational metaphors around easily, without needing to go back to each parent page to figure out what it was stuffing into what columns and why. They would know what they needed to know for the template because the passed in attributes said what they needed, not how it should be resolved. 

...IMHO.

**Comment by recena** [comment]
@jenkinsci/code-reviewers are welcome!

**Comment by recena** [comment]
I don't know way this PR is still waiting to be merged 😄

**Comment by oleg-nenashev** [comment]
👍 and 🐝 
I think that comments from @gusreiber make some sense, but fixing some low-hanging fruits in the current design is reasonable.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 2]
```diff
@@ -37,24 +37,24 @@ THE SOFTWARE.
...
+      *the use of the attribute is discouraged now.*
       plugins should now do so by inserting &lt;style> elements and/or &lt;script> elements
       in &lt;l:header/> tag.
     </st:attribute>
     <st:attribute name="permission">
       If given, this page is only made available to users that has the specified permission.
       (The permission will be checked against the "it" object.)
     </st:attribute>
+    <st:attribute name="type" use="optional">
+      Available values: two-column (by default) or one-column (full-width size).
```
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/layout.jelly]: Seems three-column layout can be easily added once we need it. Nice

Comment (recena) [inline core/src/main/resources/lib/layout/layout.jelly]: Sure, new layouts can be added through `<layout />` easily.

**Comment by recena** [comment]
I sent these two commits in order to resolve the conflicts with `master`.

**Comment by batmat** [comment]
@recena IMO it'd be preferable to just force push a rebased/amended version. In the future, having commits in the history about conflicts with master is probably gonna be less useful than having only the two original commits you pushed (modified).

**Comment by oleg-nenashev** [comment]
@batmat @recena 
I can just squash commits during the merge.

@batmat Are you fine with the PR?

**Comment by recena** [comment]
@batmat Thanks

**Comment by recena** [comment]
@oleg-nenashev I agree with squash the commits involved in this PR.

**Comment by oleg-nenashev** [comment]
@recena done

**Comment by recena** [comment]
@oleg-nenashev Thanks so much! /cc @KostyaSha

**Comment by KostyaSha** [comment]
@recena i was almost ready to sign ICLA to merge this PR.

**Comment by recena** [comment]
@KostyaSha If you have any comment or feedback about the feature implemented in this PR, please, don't hesitate to ping. I'll be glad to do improvements.

**Comment by KostyaSha** [comment]
only found that disk-usage-plugin has sidebar without any links.

**Comment by recena** [comment]
@KostyaSha Could you file a Jira issue and assign me? Meantime, I'm going to verify it.

**Comment by daniel-beck** [comment]
@KostyaSha Hasn't that always been that way?

**Comment by recena** [comment]
@KostyaSha In fact, I did not find any problem.

**Pull request merged**